python/cxx/export.cpp

```diff
@@ -27,25 +30,60 @@ void python::common::export_all(py::module& module) {
     export_ParserKeywords(module);
 }
 
-
 PYBIND11_MODULE(opmcommon_python, module) {
     python::common::export_all(module);
 }
 
-void python::common::export_all_opm_embedded(py::module& module) {
-    python::common::export_all(module);
+/**
+ * @brief Function to export several classes to the opm_embedded module that is created as a shared library.
+ * 
+ * @param module Reference to the python module.
+ */
+void python::common::export_all_opm_embedded_for_sl(py::module& module) {
+    export_Schedule_embedded(module);
+    export_EclipseState_embedded(module);
+    export_SummaryState_embedded(module);
+
+    export_Connection(module);
+    export_FieldProperties(module);
+    export_EclipseGrid(module);
+    export_Group(module);
+    export_Log(module);
+    export_ScheduleState(module);
+    export_SimulationConfig(module);
+    export_TableManager(module);
+    export_Well(module);
+
     module.attr("current_ecl_state") = std::make_shared<EclipseState>();
     module.attr("current_summary_state") = std::make_shared<SummaryState>();
     module.attr("current_schedule") = std::make_shared<Schedule>();
     module.attr("current_report_step") = 0;
     module.doc() = R"pbdoc(This is the opm_embedded module for embedding python code in PYACTION.)pbdoc";
 }
 
+/**
+ * @brief Function to export several classes to the opm_embedded module.
+ * 
+ * @param module Reference to the python module.
+ */
+void python::common::export_all_opm_embedded(py::module& module) {
+    // The additional classes are needed for the PYINPUT keyword
+    export_ParseContext(module);
+    export_Parser(module);
+    export_Deck(module);
+    export_DeckKeyword(module);
+    export_UnitSystem(module);
+    export_ParserKeywords(module);
+
+    python::common::export_all_opm_embedded_for_sl(module);
+}
+
+
 /*
   PYBIND11_MODULE create a Python of all the Python/C++ classes which are
-  generated in the python::common::export_all_opm_embedded() function in the wrapping code.
-  The same module is created as an embedded python module in PythonInterp.cpp
+  generated in the python::common::export_all_opm_embedded_for_sl() function in the wrapping code.
+  The same module (with export_all_opm_embedded) is created as an embedded python module in PythonInterp.cpp
 */
 PYBIND11_MODULE(opm_embedded, module) {
```

Comment:
I guess having more than one  `PYBIND11_MODULE()` call in a `.cpp` file will work (in this file we have two calls). At least, the jenkins build shows that it compiles. But the two produced shared objects `opmcommon_python.so` and `opm_embedded.so` will share exactly the same code (on my computer they are both 235 Mb in size) which makes me think there could be better ways to do this (avoiding generating two identical 235 Mb files).

Reply:
Good catch:
```
mblatt@frau-mahlzahn:~/src/dune/opm-2.9/opm-common/opm-parallel$ find . -name \*cpython\*.so -exec md5sum \{\} \;
211aa32976b94d190905fbd4f37e632a  ./python/opm/opmcommon_python.cpython-311-x86_64-linux-gnu.so
211aa32976b94d190905fbd4f37e632a  ./python/opm_embedded.cpython-311-x86_64-linux-gnu.so
mblatt@frau-mahlzahn:~/src/dune/opm-2.9/opm-common/opm-parallel$
```

Reply:
Can you have a look at this commit? 95c13afdf1e8704cf0f22afabd9be78b2235269a

Reply:
@lisajulia Yes, I think it solves part of the problem. After the changes proposed in the last commit, I now see that the size of `opm_embedded.so` is 10 Mb less than that of `opmcommon_python.so`. But still I guess there is over 200 Mb of common code in the two. So we might want to investigate other options to avoid this code duplication. For example, one idea could be to just keeping one of the two shared objects and instead implementing `opm_embedded` as a subclass inside the `opmcommon_python.so` shared object.

Then a PYACTION script could use for example `import opm.embedded` to import the `embedded` package from the `opm` namespace. And the `opm.embedded`  package can be created virtually like for example the `opm.io` package, see: https://github.com/OPM/opm-common/blob/master/python/opm/io/__init__.py